File: frontend/src/lib/api/segment-routing.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface Srv6Locator {
  name: string;
  prefix: string | null;
  block_len: number | null;
  node_len: number | null;
  func_bits: number | null;
  behavior_usid: boolean;
}

export interface SrInterface {
  name: string;
  hmac: string | null;
}

export interface SegmentRoutingConfig {
  locators: Srv6Locator[];
  interfaces: SrInterface[];
}

export interface SegmentRoutingCapabilities {
  version: string;
  version_info: {
    is_1_4: boolean;
    is_1_5: boolean;
    /**
     * True on VyOS 1.4, where FRR rejects any in-place modification of an
     * existing segment-routing tree: edits must delete the tree and recreate
     * it in two separate commits.
     */
    modify_requires_recreate: boolean;
  };
  features: {
    locators: { supported: boolean; description: string };
    interface_srv6: { supported: boolean; description: string };
  };
  instance_name?: string;
  instance_id?: string;
}

export interface SegmentRoutingBatchOperation {
  op: string;
  value?: string;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

// ============================================================================
// API Service
// ============================================================================

class SegmentRoutingService {
  async getCapabilities(): Promise<SegmentRoutingCapabilities> {
    return apiClient.get<SegmentRoutingCapabilities>("/vyos/segment-routing/capabilities");
  }

  async getConfig(refresh = false): Promise<SegmentRoutingConfig> {
    return apiClient.get<SegmentRoutingConfig>("/vyos/segment-routing/config", {
      refresh: refresh.toString(),
    });
  }

  private async batch(ops: SegmentRoutingBatchOperation[]): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/segment-routing/batch", {
      operations: ops,
    });
    if (!result.success) throw new Error(result.error || "Segment Routing operation failed");
    return result;
  }

  private locatorOps(locator: Srv6Locator): SegmentRoutingBatchOperation[] {
    const ops: SegmentRoutingBatchOperation[] = [];
    if (locator.prefix) {
      ops.push({ op: "set_locator_prefix", value: `${locator.name},${locator.prefix}` });
    }
    if (locator.block_len != null) {
      ops.push({ op: "set_locator_block_len", value: `${locator.name},${locator.block_len}` });
    }
    if (locator.node_len != null) {
      ops.push({ op: "set_locator_node_len", value: `${locator.name},${locator.node_len}` });
    }
    if (locator.func_bits != null) {
      ops.push({ op: "set_locator_func_bits", value: `${locator.name},${locator.func_bits}` });
    }
    if (locator.behavior_usid) {
      ops.push({ op: "set_locator_behavior_usid", value: locator.name });
    }
    return ops;
  }

  /** All operations that reproduce a full desired config from an empty tree. */
  private fullConfigOps(config: SegmentRoutingConfig): SegmentRoutingBatchOperation[] {
    const ops: SegmentRoutingBatchOperation[] = [];
    // Interfaces first: VyOS rejects a locator commit unless at least one
    // interface has SRv6 enabled in the same commit.
    for (const iface of config.interfaces) {
      if (iface.hmac) {
        ops.push({ op: "set_interface_hmac", value: `${iface.name},${iface.hmac}` });
      } else {
        ops.push({ op: "set_interface_srv6", value: iface.name });
      }
    }
    for (const locator of config.locators) {
      ops.push(...this.locatorOps(locator));
    }
    return ops;
  }

  /**
   * Create a locator. When the config has no SRv6-enabled interface yet,
   * pass enableInterface to bundle the interface enable into the same
   * commit — VyOS rejects a locator without one.
   */
  async createLocator(locator: Srv6Locator, enableInterface?: string): Promise<VyOSResponse> {
    const ops: SegmentRoutingBatchOperation[] = [];
    if (enableInterface) {
      ops.push({ op: "set_interface_srv6", value: enableInterface });
    }
    ops.push(...this.locatorOps(locator));
    return this.batch(ops);
  }

  /** In-place locator update (works on VyOS 1.5/rolling). */
  async updateLocator(original: Srv6Locator, updated: Srv6Locator): Promise<VyOSResponse> {
    const ops: SegmentRoutingBatchOperation[] = [];
    const name = original.name;

    if (original.prefix !== updated.prefix && updated.prefix) {
      ops.push({ op: "set_locator_prefix", value: `${name},${updated.prefix}` });
    }
    if (original.block_len !== updated.block_len) {
      if (updated.block_len != null) {
        ops.push({ op: "set_locator_block_len", value: `${name},${updated.block_len}` });
      } else {
        ops.push({ op: "delete_locator_block_len", value: name });
      }
    }
    if (original.node_len !== updated.node_len) {
      if (updated.node_len != null) {
        ops.push({ op: "set_locator_node_len", value: `${name},${updated.node_len}` });
      } else {
        ops.push({ op: "delete_locator_node_len", value: name });
      }
    }
    if (original.func_bits !== updated.func_bits) {
      if (updated.func_bits != null) {
        ops.push({ op: "set_locator_func_bits", value: `${name},${updated.func_bits}` });
      } else {
        ops.push({ op: "delete_locator_func_bits", value: name });
      }
    }
    if (original.behavior_usid !== updated.behavior_usid) {
      if (updated.behavior_usid) {
        ops.push({ op: "set_locator_behavior_usid", value: name });
      } else {
        ops.push({ op: "delete_locator_behavior_usid", value: name });
      }
    }

    if (ops.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }
    return this.batch(ops);
  }

  /** In-place locator delete (works on VyOS 1.5/rolling). */
  async deleteLocator(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_locator", value: name }]);
  }

  /**
   * VyOS 1.4 mutation path: FRR rejects in-place changes to an existing
   * segment-routing tree, so apply the desired end state by deleting the
   * whole tree (commit 1) and recreating it (commit 2). If the second
   * commit fails, the tree is left empty — callers must warn the user.
   */
  async applyViaRecreate(desired: SegmentRoutingConfig): Promise<VyOSResponse> {
    await this.batch([{ op: "delete_segment_routing" }]);
    const ops = this.fullConfigOps(desired);
    if (ops.length === 0) {
      return { success: true, data: { message: "Segment Routing configuration removed" } };
    }
    return this.batch(ops);
  }
}

export const segmentRoutingService = new SegmentRoutingService();

File: backend/routers/segment_routing/segment_routing.py
"""Segment Routing Protocol Router.

API endpoints for managing VyOS Segment Routing (SRv6) configuration:
locators and per-interface SRv6 packet acceptance. The configuration tree
is identical on VyOS 1.4 and 1.5.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import SegmentRoutingBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/segment-routing", tags=["segment-routing"])


# ============================================================================
# Pydantic Models
# ============================================================================


class Srv6Locator(BaseModel):
    name: str
    prefix: Optional[str] = None
    block_len: Optional[int] = None
    node_len: Optional[int] = None
    func_bits: Optional[int] = None
    behavior_usid: bool = False


class SrInterface(BaseModel):
    name: str
    hmac: Optional[str] = None


class SegmentRoutingConfig(BaseModel):
    locators: List[Srv6Locator] = []
    interfaces: List[SrInterface] = []


class SegmentRoutingBatchOperation(BaseModel):
    op: str = Field(..., description="Builder method name")
    value: Optional[str] = Field(None, description="Comma-separated arguments")


class SegmentRoutingBatchRequest(BaseModel):
    operations: List[SegmentRoutingBatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_segment_routing_capabilities(request: Request):
    """Return Segment Routing feature capabilities based on the device VyOS version.

    version_info.modify_requires_recreate is true on VyOS 1.4, where FRR
    rejects any modification of an existing segment-routing tree: clients
    must delete the tree and recreate it in two separate batches to edit.
    """
    await require_read_permission(request, FeatureGroup.SEGMENT_ROUTING)

    try:
        service = get_session_vyos_service(request)
        builder = SegmentRoutingBatchBuilder(version=service.get_version())
        capabilities = builder.get_capabilities()

        if hasattr(request.state, "instance") and request.state.instance:
            capabilities["instance_name"] = request.state.instance.get("name")
            capabilities["instance_id"] = request.state.instance.get("id")

        return capabilities
    except Exception:
        logger.exception("Unhandled error in get_segment_routing_capabilities")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=SegmentRoutingConfig)
async def get_segment_routing_config(http_request: Request, refresh: bool = False):
    """Return the full Segment Routing configuration in a normalized format."""
    await require_read_permission(http_request, FeatureGroup.SEGMENT_ROUTING)

    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        sr_raw = full_config.get("protocols", {}).get("segment-routing", {})

        if not sr_raw:
            return SegmentRoutingConfig()

        return SegmentRoutingConfig(
            locators=_parse_locators(sr_raw.get("srv6", {}).get("locator", {})),
            interfaces=_parse_interfaces(sr_raw.get("interface", {})),
        )
    except Exception:
        logger.exception("Unhandled error in get_segment_routing_config")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Config Parsers
# ============================================================================


def _safe_int(value) -> Optional[int]:
    if value is None:
        return None
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def _parse_locators(raw: dict) -> List[Srv6Locator]:
    if not raw:
        return []

    locators = []
    for name, cfg in raw.items():
        if cfg is None:
            cfg = {}

        locators.append(Srv6Locator(
            name=name,
            prefix=cfg.get("prefix"),
            block_len=_safe_int(cfg.get("block-len")),
            node_len=_safe_int(cfg.get("node-len")),
            func_bits=_safe_int(cfg.get("func-bits")),
            behavior_usid="behavior-usid" in cfg,
        ))

    return locators


def _parse_interfaces(raw: dict) -> List[SrInterface]:
    if not raw:
        return []

    interfaces = []
    for name, cfg in raw.items():
        if cfg is None:
            cfg = {}

        srv6 = cfg.get("srv6", {}) or {}
        interfaces.append(SrInterface(
            name=name,
            hmac=srv6.get("hmac"),
        ))

    return interfaces


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def segment_routing_batch_configure(http_request: Request, body: SegmentRoutingBatchRequest):
    """Execute a batch of Segment Routing configuration operations atomically."""
    await require_write_permission(http_request, FeatureGroup.SEGMENT_ROUTING)

    try:
        service = get_session_vyos_service(http_request)
        builder = SegmentRoutingBatchBuilder(version=service.get_version())

        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 0:
                method()
            elif len(params) == 1:
                if operation.value is not None:
                    method(operation.value)
            elif len(params) == 2 and operation.value is not None:
                values = operation.value.split(",", 1)
                if len(values) == 2:
                    method(values[0], values[1])
                else:
                    method(operation.value, "")

        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Segment Routing configuration updated"},
            error=response.error if response.error else None,
        )
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {str(e)}")
    except Exception:
        logger.exception("Unhandled error in segment_routing_batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")
